### ISSUE TITLE
Testing support in ionic2-app-base

### ISSUE THREAD

**Issue by brandyscarney** [ionic-team/ionic-app-scripts#687]
_From @BugsBunnyBR on February 23, 2016 21:55_

#### Short description of the problem:

the project ionic2-app-base does not provide a easy way to create tests for the project.
#### What behavior are you expecting?

I would like to have an easy way to integrate tests to my project.

**Ionic Version**:  2.x

@lathonez and I would like to add support to testing in ionic2-app-base, he already got ionic2 projects with tests in his [clicker](https://github.com/lathonez/clicker) repo, so we are willing to integrate this to the app-base, ideally having a unified build process for testing and development.
We also would like to hear about ideias e patterns the community/Ionic team would suggest.

_Copied from original issue: driftyco/ionic#5579_

**Comment by brandyscarney** [comment]
_From @lathonez on February 23, 2016 22:3_

At this point some action is better than none. There's certainly a better way of testing than what I've implemented in [Clicker](https://github.com/lathonez/clicker), but if we have a starting point we can evolve it with the community as the framework grows.

We've [forked the base app](https://github.com/BugsBunnyBR/ionic2-app-base) and are going to put some work into getting the testing set up there.

see also https://github.com/driftyco/ionic/issues/5331

**Comment by brandyscarney** [comment]
_From @tlancina on February 24, 2016 16:41_

Thanks guys! Really appreciate the discussion going on in https://github.com/driftyco/ionic/issues/5331.  Just to give you a heads up, we will be pulling much of the build out into a gulpfile, which should give more granular control of the various tasks, and would also allow the easy addition of testing tasks.  Also just a warning we may be moving off of webpack!

The starters are going through quite a few changes at the moment, so I can't give testing much attention at this exact moment, but a PR would be absolutely huge and I think a great starting point for discussion/getting things up and running in practice.  Instead of app-base, the [conference app](https://github.com/driftyco/ionic-conference-app) 
may be easier since it's already a fully fledged app and doesn't require copying any templates.  If it's rough no worries, we can iterate as we find better solutions.

Pattern-wise I think having the tests be alongside the components they are describing is what we have seen as the standard, other than that I'm not aware of any strongly held conventions.  Please ping me if you have any build questions/issues, thanks!

**Comment by brandyscarney** [comment]
_From @BugsBunnyBR on February 24, 2016 17:19_

Thank you @tlancina for this huge heads up!

gulpfile ftw! This will save us a lot of trouble trying to make a test webpack file.

We can add testing to the conference app, but to make testing really easier to integrate in a ionic2 project or to new ionic2 projects, app-base seems to be the right place.
Our plan, at the moment, is to fork app-base and add testing support to the typescript branch with lint optional. All documented.

We are going with tests alongside with components.

If the ionic team thinks that keep a typescript branch without support to testing is something desirable...
One thing I think could be done is to have the support in ionic-cli, something like `ionic start project --v2 --ts --test`, the `--test` would trigger the download of a specific branch with testing enabled.

**Comment by brandyscarney** [comment]
_From @tlancina on February 24, 2016 17:28_

@BugsBunnyBR I had the same thought regarding a flag as well, only because of the additional install time necessary for the various testing related packages, which may not always be desirable.

I think tests are a good default, I just want to make sure things are as easy and flexible as possible.  One idea we've toyed around with is allowing the start command to ask a series of questions regarding project setup - by default we download what we think to be the best presets, but if you are a power user you could specify things like which bundler to use, TypeScript vs. JavaScript, tests, etc (similar to yeoman).  For now, I don't think we need to worry about it, but definitely something to think about going forward.

Also regarding the conference app vs. app-base, they get updated at the same time, I just find developing against the conference app to be easier.

**Comment by brandyscarney** [comment]
_From @BugsBunnyBR on February 24, 2016 18:6_

@tlancina , if tests are a good default, the additional install time seems to pay off.
but, well...
So, how about that. We would make PRs to app-base (in a given branch) and apply the diff in conference-app typescript branch files.
- Keeping test support in a given app-base branch would make ionic team worry free and the app-base would continue clean. When/if the ionic-cli become able to read the `--test` the work would be just swap branchs...
- Applying the diff into conference-app would give a public test case to the testing support.

**Comment by brandyscarney** [comment]
_From @lathonez on February 24, 2016 20:49_

@tlancina really helpful advice particularly re gulp vs webpack, thanks for your time on this.

**Comment by brandyscarney** [comment]
_From @ciekawy on February 25, 2016 12:56_

As the discussion and existing initiatives concern unit testing in general I wonder if there are any concepts or places to look for e2e ideas for Ionic 2?

**Comment by brandyscarney** [comment]
_From @lathonez on February 25, 2016 20:32_

https://github.com/BugsBunnyBR/ionic2-app-base/issues/4

**Comment by brandyscarney** [comment]
_From @BugsBunnyBR on February 29, 2016 17:2_

https://forum.ionicframework.com/t/attention-upcoming-ionic-2-starter-changes/45058

**Comment by brandyscarney** [comment]
_From @BugsBunnyBR on March 23, 2016 18:55_

@tlancina : https://github.com/driftyco/ionic-conference-app/pull/168 , please, take a look, some feedback would be nice.

**Comment by lordgreg** [comment]
Hi. Are there any updates on conference app, app scripts or ionic itself about at least implementing karma and unit tests for start?

Thank you.

**Comment by fifafu** [comment]
I would also love to have unit & e2e testing integrated in the Ionic App Scripts. I think it's the only thing that's still missing and it would be great if it'd be integrated.
Maybe just use the setup from the Angular CLI? AFAIK testing has been working quite well with their setup.

**Comment by lordgreg** [comment]
I'd love to hear anything about the issue from @adamdbradley or @danbucholtz since the relate MR in ionic-conference-app from @BuggsBunnyBR was correct but wow.. the dependencies you need. ohh boy.

**Comment by danbucholtz** [comment]
@lordgreg,

We will add testing support at some point. We're locked in on shrinking the bundle size and improving start-up time right now.

Thanks,
Dan

**Comment by lordgreg** [comment]
Thank you for your response @danbucholtz.

Good luck in achieving goals :)

**Comment by rapropos** [comment]
I don't know if others are having the same problem, but I am finding that [lack of support for TypeScript 2.1](https://github.com/driftyco/ionic-app-scripts/issues/197) is a fatal blocker for me when trying to make a FrankenProject combining angular-cli's support for unit testing with ionic-app-scripts. If I generate the project with angular-cli and graft the ionic stuff onto it, testing seems to build but `ionic build` fails due to complaints about initializers in ambient context when referring to environment files. According to my searches for similar errors in other, this apparently goes away when TS is upgraded to 2.1. If I go the other way around and try to drop angular-cli infrastructure into an ionic-generated project, testing won't build. Downgrading an angular-cli built project to use TS 2.0 breaks the testing build.

TL;DR: I think at least attempting to allow use of TypeScript 2.1 would be a big step forward in terms of helping people set up unit testing for Ionic projects.

**Comment by lordgreg** [comment]
@rapropos I am not sure how unit or e2e have anything to do with Typescript versioning. For that, I'd open another issue. This is more or less a topic about the possibility to use karma/protractor within ionic2 Project.

👍

**Comment by lathonez** [comment]
@rapropos - https://github.com/lathonez/clicker is running @angular/cli rc0, @angular/* 2.4 and @ionic 2.1. Updated yesterday.

**Comment by danbucholtz** [comment]
@rapropos,

We can't until we support Angular 4. Angular 2.x doesn't work with TS 2.1.

We are very anxious for this. In your own project, feel free to give ng 4 RC a shot.

Thanks,
Dan

**Comment by rapropos** [comment]
After examining @lathonez's clicker project, I managed to figure out that the environment files that angular-cli creates (which were the source of the <TS2.1 errors) are not 100% needed when ordinarily building with ionic-app-scripts. Deleting them allowed me to make some progress.

**Comment by danielsogl** [comment]
@brandyscarney can you close it?
